### PR TITLE
CRM-21403: Membership advanced search result fix when Current Member Input Field is NO

### PR DESCRIPTION
**Pull request by roliod**
Overview
----------------------------------------
Membership advanced search no longer shows a contact with an active membership when the 'Current Member' input field is 'No'. 

Before
----------------------------------------
If a contact has multiple memberships where one is no longer active and the other is. 

screenshot 2018-04-19 21 38 21

The contact appears in the Membership advanced search result when the 'Current Member' input field is checked as 'No'.

screenshot 2018-04-19 21 37 54
screenshot 2018-04-19 21 38 11

After
----------------------------------------
If the 'Current member' input field is checked as No. For as long as one of the contacts multiple memberships is active, the contact does not appear in the search result.

Technical Details
----------------------------------------
Within the loop of the contact result, there is a check to see if the specific contact has an active membership.

---

 * [CRM-21403: 'Current Member' field in searches does not perform as expected](https://issues.civicrm.org/jira/browse/CRM-21403)

https://lab.civicrm.org/dev/core/issues/132

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by seamuslee001** [comment]
Jenkins ok to test

**Comment by eileenmcnaughton** [comment]
So I have a few thoughts here

Most importantly I am unsure whether this is a straight forward bug fix or something where people might have different opinions about correct behaviour - some combo of @Stoob @petednz @davejenx @lcdservices @agh1 @ErikHommel @joannechester might be able to confirm here...

Implementation comments...

1) we should be able to avoid getting from POST - seems like $form->_object->_queryParams has the data -

2) filtering in getRows might mess with the paging ? If the row-count is different to the foundResults
The actual place the query is done is here - & perhaps a subquery makes sense there - not sure
 https://github.com/JKingsnorth/civicrm-core/blob/8c9251b398a29d9b82996ae21825c9d6478fa576/CRM/Member/BAO/Query.php#L265 might be the place to try to 

3) the test is not actually testing the same path because the api uses a different function to get results :-(

**Comment by lcdservices** [comment]
I'm not sure I agree with this implementation. It's admittedly a fuzzy area from a spec standpoint, as different orgs may interpret their membership scenario in different ways. But here are my thoughts:

- advanced search, when choosing to display contacts (default behavior), is interacting with related records (memberships) and returning distinct contacts. so there's a delicate interplay there between what your expected results would be when viewing the related records vs. when viewing the contact.
- so the question I ask is -- if we chose to display membership records instead of contacts, what would be the expected result. in this case, if I search for memberships where the membership is not current, I would expect to get a list of all membership meeting that criteria, regardless of whether the contact has other memberships that _are_ current.
- therefore, if searching and displaying memberships would yield a result for this search, I expect displaying contacts for the same search would return a result.

To put it another way -- searching for contacts with a non-current membership and searching for contacts with at least one current membership are not identical searches. If I were to do the latter, I would probably create a smart group of contacts with a current membership, then use the include/exclude custom search to identify contacts not in that group (the actual requirements may be a bit more nuanced -- such as finding everyone with a membership, regardless of status, and excluding those with a current membership; but the idea is the same).

In my current thinking, I'd lean away from this change.

**Comment by Stoob** [comment]
I agree with Brian on this issue specifically, and he is correct.  A 
person with multiple membership status will show up in multiple results 
because this is correct.  Generally, if a feature has been working the 
same way for a decade, there is a precedent and reason for the software 
working the way it does.  A person may not agree with that, but it is 
the status quo.  I hope in the future there is more communication about 
these issues before the work is done to submit a patch.  It would save 
everyone including the submitter their time.

Also: this change would break many smart groups people have built.

Workaround: if you wish to return the membership entity(s) not the 
contact you can use the "Display Results As...".  This would/should 
return 3 rows for 3 memberships for the single contact listed in the 
example.

**Comment by joannechester** [comment]
I agree that we shouldn't change the way that search criterion works, However I think we should change the label from "Current member?" to something that _may_ be more meaningful to users ( or at least easier to explain to them), perhaps "Membership is current?" ( And I would dearly love to change 'current' to, say, 'valid'.)

Having said that, I can see that having a tick box option to let them find someone who has no active membership would be a useful feature for users. Is there any capacity to implement this logic in the memberships accordian but have it only appear there for an advanced search when the results are being displayed as contacts?

**Comment by roliod** [comment]
@Stoob @eileenmcnaughton @lcdservices do you agree with @joannechester ?

**Comment by lcdservices** [comment]
I'm fine with changing the terminology to "Membership is current" -- or maybe just "Current membership" (either one shifts the perception of status to the membership record, away from the contact, which is where I agree there could be confusion). I can see value in that clarification.

Re: the second suggestion -- if I understand correctly, @joannechester is suggesting adding what @roliod has proposed as a new option (preserving the current behavior as-is). That sounds fine to me. We may want to consider reworking the existing field so that it's a list of options rather than add a new field. I'm not sure how we'd label it, but the options could be something like: 

- Current memberships only
- At least one current membership [new option]
- No current memberships

We'd probably want some well crafted help text to clarify the distinctions.

I think it's important to retain the term "current" as this filter directly relates to the option in the membership status rules to flag a status as current.

**Comment by joannechester** [comment]
To clarify, when I said I would like to change 'Current' to 'Valid' I was meaning a change not just be to the filter but also to the status rules setup. So instead of:

> Current Membership? | Should this status be considered a  current membership in good standing. EXAMPLE: New, Current and Grace  could all be considered 'current'.


It would be

> Valid Membership? | Should this status be considered a membership in good standing. EXAMPLE: New, Current and Grace  could all be considered 'valid'.

If the existing field was to be rewoked to include a third option my suggestions for options would be:

-  Any current memberships 
-  At least one membership is current [new option]
-  Any non-current memberships

However, I think it would be cleaner to have a new tick box filter with the label "At least one current membership".

**Comment by davejenx** [comment]
I agree that users could reasonably have different expectations of what results should be displayed when the 'Current Member' input field is 'No' and results are displayed as contacts. I'd guess among users not steeped in Civi's traditions, a majority would expect it to list only contacts who don't have a current membership, though there's certainly room for debate particularly if no membership type is specified. Even if a membership type is specified, some organisations, especially where memberships have been imported from other systems, have multiple membership records of the same type per contact, e.g. two expired and one current. In the latter case, users typically expect 'Current Member' = 'No' to show only contacts with no current membership (of the specified type).

I agree with others here that rather than changing the current debatable behaviour to different debatable behaviour, it would be better to provide clear and distinct options. The options need a bit more thought in my view, taking into account how they would interact with both the membership type field and with display as contacts vs memberships, in use cases that may involve contacts with multiple memberships of both the same and different types.

Andrew makes some very good points about is_current_member. Although I take the point about not wanting to further entrench a problematic field, it's a wider discussion and I'm not sure it needs to block progress on the current issue. If there is subsequent agreement to replace is_current_member with e.g. in_good_standing, renew_without_lapse and eligible_for_member_benefits, perhaps we could amend the solution to the current issue accordingly?

**Comment by eileenmcnaughton** [comment]
I'm going to close this since it seems fairly clear that this is a behaviour change for which we have not yet agreed a spec for. Closing a PR does not lose any of the code that has been done or any of the discussion. It does mean it is no longer seen as being in the queue for code review (which is appropriate while the proposal is under discussion).

https://lab.civicrm.org/dev/core/issues/132 for discussion.
